src/Hedgehog/Gen.fs

```diff
@@ -2,18 +2,36 @@
 
 open System
 
+type GenConfig<'a> = private {
+    Formatter : 'a -> string
+}
+
 /// A generator for values and shrink trees of type 'a.
 [<Struct>]
-type Gen<'a> =
-    | Gen of Random<Tree<'a>>
+type Gen<'a> = private {
+    Random : Random<Tree<'a>>
+    Config : GenConfig<'a>
+}
 
 module Gen =
 
-    let ofRandom (r : Random<Tree<'a>>) : Gen<'a> =
-        Gen r
+    let format (a : 'a) (gen : Gen<'a>) : string =
+        gen.Config.Formatter a
+
+    let withFormatter (formatter : 'a -> string) (gen : Gen<'a>) : Gen<'a> =
+        {
+            gen with Config = { gen.Config with Formatter = formatter }
+        }
+
+    let ofRandom (r : Random<Tree<'a>>) : Gen<'a> = {
```

Comment:
As noted in #346, this needs to be traced carefully so we don't end up overriding the configuration.